lib/python/pygrass/utils.py

```diff
@@ -389,6 +389,426 @@ def table_exist(cursor, table_name):
+                    break
+                except:
+                    continue
+            if not dtype:
+                dtype = np_array_view[:, col_idx].dtype
+            formats.append((dtype))
+
+    # Generate a list of tuples with column names and formats for the array
+    if not names and formats:
+        dtype = np.dtype(
+            [
+                (
+                    "{}{}".format(column_prefix, idx)
+                    if not np_array.dtype.names
+                    else np_array.dtype.names[idx],
+                    np_format,
+                )
+                for idx, np_format in enumerate(formats)
+            ]
+        )
+    elif names and formats:
+        dtype = np.dtype(
+            [(names[idx], np_format,) for idx, np_format in enumerate(formats)]
+        )
+    elif names and not formats:
+        dtype = np.dtype([(np_name, names[idx],) for idx, np_name in enumerate(names)])
+    else:
+        dtype = None
+
+    # Start with unstructured array
+    if not np_array.dtype.names or update_formats or formats:
+        structured_array = rfn.unstructured_to_structured(np_array_view, dtype)
+    else:
+        structured_array = np_array
+
+    # Generate a list of SQL data types for columns
+    columns = []
+    for col in structured_array.dtype.names:
+        type_code = structured_array[col].dtype.num
+        columns.append("{} {}".format(col, dtype_to_sql[dbdriver][type_code]))
+
+    # Define initial SQL strings
+    create_sql = "CREATE TABLE {} ({});".format(table, ", ".join(columns))
+
+    # Execute SQL code
+    with connection:
+        cur = connection.cursor()
+        if overwrite:
+            drop_sql = "DROP TABLE IF EXISTS {};".format(table)
+            cur.execute(drop_sql)
+        # Create table
+        cur.execute(create_sql)
+
+        # Insert data
+        if dbdriver == "sqlite":
+            insert_sql = "INSERT INTO {}({}) VALUES({});".format(
+                table,
+                ", ".join(structured_array.dtype.names),
+                ",".join(["?"] * len(structured_array.dtype.names)),
+            )
+            cur.executemany(insert_sql, structured_array.tolist())
+        elif dbdriver == "pg":
+            # For arrays that do not contain objects or binary data, they could be loaded using the copy statement
+            if set(
+                [np.dtype(descr[1]).num for descr in structured_array.dtype.descr]
+            ).isdisjoint(set([13, 14, 15, 16, 17, 20])):
+                np_array_txt = BytesIO()
+                np.savetxt(np_array_txt, structured_array, delimiter="\t", fmt="%s")
+                np_array_txt.seek(0)
+                cur.copy_from(np_array_txt, table)
+            else:
+                insert_sql = "INSERT INTO {}({}) VALUES %s;".format(
+                    table,
+                    ", ".join(structured_array.dtype.names),
+                    ",".join(["?"] * len(structured_array.dtype.names)),
```

Comment:
Use named placeholders such as `{table}` and `table=` when there is more than one item to avoid confusion. Here seems to be some with `%s` and the third argument.